File: BOJ(Java 공부용)/7562.java
import java.io.*;
import java.util.ArrayDeque;
import java.util.StringTokenizer;
public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    static int[] gox = {-1, -1, 1, 1, -2, -2, 2, 2};
    static int[] goy = {-2, 2, -2, 2, 1, -1, 1, -1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            int n = Integer.parseInt(br.readLine());
            boolean[][] visited = new boolean[n][n];
            StringTokenizer st = new StringTokenizer(br.readLine());
            int sx = Integer.parseInt(st.nextToken()), sy = Integer.parseInt(st.nextToken());
            st = new StringTokenizer(br.readLine());
            int ex = Integer.parseInt(st.nextToken()), ey = Integer.parseInt(st.nextToken());
            bw.write(bfs(sx, sy, ex, ey, visited, n) + "\n");
        }
        bw.flush();
        bw.close();
    }

    private static int bfs(int sx, int sy, int ex, int ey, boolean[][] visited, int n) {
        ArrayDeque<p> q = new ArrayDeque<>();
        q.add(new p(sx, sy, 0));
        visited[sx][sy] = true;
        int temp = 0;
        while (!q.isEmpty()) {
            int x = q.peek().x;
            int y = q.peek().y;
            int dist = q.peek().dist;
            q.poll();
            if (x == ex && y == ey) {
                temp = dist;
                break;
            }
            for (int i = 0; i < 8; i++) {
                int nx = x + gox[i], ny = y + goy[i];
                if (nx < 0 || nx >= n || ny < 0 || ny >= n || visited[nx][ny]) continue;
                q.add(new p(nx, ny, dist + 1));
                visited[nx][ny] = true;
            }
        }
        return temp;
    }

    private static class p {
        int x, y, dist;

        private p(int x, int y, int dist) {
            this.x = x;
            this.y = y;
            this.dist = dist;
        }
    }
}

File: BOJ(Kotlin)/7562.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

fun main(args: Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    val t = br.readLine().toInt()
    for (i in 0 until t) {
        val l = br.readLine().toInt()
        var st = StringTokenizer(br.readLine())
        val sx = st.nextToken().toInt()
        val sy = st.nextToken().toInt()
        st = StringTokenizer(br.readLine())
        val ex = st.nextToken().toInt()
        val ey = st.nextToken().toInt()
        bw.write(bfs(l,sx,sy,ex,ey).toString()+"\n")
    }
    bw.flush()
    bw.close()
}
fun bfs(l : Int, sx : Int, sy : Int, ex : Int, ey : Int) : Int {
    val visited = Array(l){BooleanArray(l)}
    val dq = ArrayDeque<Night>()
    dq.add(Night(sx,sy,0))
    val gox = intArrayOf(-1,-1,-2,-2,1,1,2,2)
    val goy = intArrayOf(-2,2,-1,1,-2,2,-1,1)
    var ans = 0
    while(!dq.isEmpty()) {
        val x = dq.first.x
        val y = dq.first.y
        val move = dq.first.move
        dq.pollFirst()
        if (x == ex && y == ey) {
            ans = move
            break
        }
        for (i in 0..7) {
            val nx = x + gox[i]
            val ny = y+goy[i]
            if (nx < 0 || nx>= l || ny<0 || ny>=l || visited[nx][ny]) continue
            dq.add(Night(nx,ny,move+1))
            visited[nx][ny] = true
        }
    }

    return ans
}

class Night(val x : Int, val y : Int, val move : Int){}